### PR TITLE
Simplify DOSBox Staging version handling

### PR DESCRIPTION
**Pull request by johnnovak**
# Description

## The current way 😐 

The current DOSBox Staging version handling is a nightmare. It has too many moving parts:

- it relies on Git tags,
- manual tinkering with `git describe --abbrev` on the release branch,
- requires us to force-fetch the whole (yes, the _whole_!) Git history with the `git fetch --prune --unshallow` plus `git fetch --all --tags --force` hack/workaround instead of relying on the checkout action's default behaviour (only do a shallow checkout of `HEAD`),
- has some weird conditional logic to not make the Git hash available in release builds (why not?),
- version retrieval is inlined in workflows in slightly different ways (lots of room for mismatches and bugs),
- there are 3-4 different ways to fetch/inject/replace the various version strings "flavours" in various templates (I stopped counting; it's different for Meso,. MVSC, MSYS2, etc.)
- ...

Ok, I think you get the picture. Bumping the version number for the next release should be a **5 second job!** Yet I've wasted about _a day_ in total massaging this overengineered machinery over the last years or so, and I've had enough...

I shit you not; getting all the stars align to get the version to show up correctly in all builds is the singular part of whole the release process that I dread the most!

But, alas, no longer! 😎 🔥 🪓 

## The new way 🥳 

- The version string is now defined _manually_ in `include/version.h`. E.g., 

	```c
	#define DOSBOX_VERSION "0.82.0-alpha"
	```
  That's it! Total control, and no more second-guessing yourself and re-running the workflows 3-5 times to get it right.

- No more reliance on tags to determine the version string means we don't need to fetch the whole history with hacks/workarounds which should be more efficient and speed the build workflows up a bit.

- The current Git hash is retrieved with `git rev-parse` which is 100% reliable. 

- Every workflow, script, Meson, etc. uses `scripts/get-version.sh` to retrieve the manually set version, the Git hash, or the combined version plus Git hash. This script is the single source of truth---no duplication anywhere anymore.

- The Git hash is always available at runtime, and it's printed out in parenthesis. No weird conditional trickery that it's only printed out in debug builds and alpha builds, but not release builds, etc.

- Release artifact names now contain the version and follow a consistent pattern:

  - `dosbox-staging-linux-x86_64-0.82.0-alpha-4ff75`
  - `dosbox-staging-windows-x64-0.82.0-alpha-4ff75`
  - `dosbox-staging-macOS-universal-0.82.0-alpha-4ff75`

- This allows us the remove that `publish_additional_artifacts` job from the build workflows which mainly existed to pass the version string to the JavaScript in the development builds page by uploading an extra changelog file. I've modified the JavaScript to extract the version directly from the release artifact names.

- Removed the ClamAV virus scan from the Linux workflow—we decided a while ago to do away with the antivirus checks in all workflows as it's a bit pointless.

## Future work

If you build locally on Windows with MSVC, the Git hash won't be injected (it will just remain set to `git`, which is the placeholder). I might improve this later, or anyone can do it who's annoyed by this (hint: you can define pre-hooks in the MSVC project for the build actions).

# Manual testing

## Development builds page

To prove the changes end-to-end, I ran the website locally with `mkdocs serve` and hacked the JS to point to the PR branch:

```javascript
let filter_branch = "jn/simpler-version-handling"
```

Result:

image

If something goes wrong, the user now at least gets an error message instead of just dumping the HTTP status code on them matter-of-factly... 😐 

image

Clicking on the links took me to the artifacts:

### Linux

image

### macOS

image

### Windows

image

## Runtime tests

This is set in `include/version.sh`:

image

And this is what `get-version.sh` gives us:

image

---

Only pasting the macOS results—I've checked the Windows build, too, and the results are the same.

image

image

image

image

image

image

image

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [ ] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

**Comment by weirddan455** [comment]
Looks like a welcome change.  I'll review in detail when I have some time.

**Review by japsmits — APPROVED** [review]
One version to rule them all. Good work!

File: meson.build [hunk 1]
```diff
@@ -1,8 +1,10 @@
 project(
-    'dosbox-staging',
-    'c',
-    'cpp',
-    version: '0.82.0-alpha',
+    'dosbox-staging', 'c', 'cpp',
+
+    version: run_command(
+      './scripts/get-version.sh', 'version', check: true,
+    ).stdout().strip(),
```
Comment (weirddan455) [inline meson.build:6]: I can see this giving us issues but I guess we can deal with them as they come up.

1. Portability.  I would like to try at some point to get Meson running with a native Windows toolchain.  Would be straight-forward to write a Windows .bat file worse comes to worst.
2. This assumes you're running the Meson command from the project's root directory.  I can foresee CI breaking here if we make certain changes.  Suggest using `meson.project_source_root()` to be more robust (I think this is the correct one, there are several variations).

File: .github/workflows/linux.yml [hunk 2]
```diff
@@ -349,31 +349,6 @@ jobs:
...
-            clamdb-linux-${{ steps.prep-clamdb.outputs.yesterday }}-1
-
-      - name: Clam AV scan
-        run: |
-          set -x
-          sudo apt-get install clamav
-          sudo systemctl stop clamav-freshclam
-          sudo sed -i 's/30/20000/g' /etc/clamav/freshclam.conf
-          sudo freshclam --foreground
-          clamscan --heuristic-scan-precedence=yes --recursive --infected .
```
Comment (weirddan455) [inline .github/workflows/linux.yml:375]: I didn't know we were doing this to begin with.  No objections to removing it but curious why you chose to do so in this PR.  Was it causing CI problems?

File: .github/workflows/linux.yml [hunk 3]
```diff
@@ -358,34 +358,3 @@ jobs:
...
           path: dosbox-staging-linux-${{ env.VERSION }}.tar.xz
           overwrite: true
-
-
-  # This job exists only to publish an artifact with version info when building
-  # from main branch, so snapshot build version will be visible on:
-  # https://dosbox-staging.github.io/downloads/devel/
-  #
-  publish_additional_artifacts:
-    name: Publish additional artifacts
```
Comment (weirddan455) [inline .github/workflows/linux.yml:393]: So this whole CI mess was just to generate a version string for the website?  Good riddance.

File: website/docs/releases/development-builds.md [hunk 4]
```diff
@@ -5,132 +5,157 @@ hide:
...
-  version_el.textContent = '';
+function set_build_version(gh_api_artifacts, os_name) {
+  fetch(gh_api_artifacts, { method: "GET", headers: headers })
+    .then(response => {
+      if (response.status !== 200)
+        return
+
+      response.json().then(data => {
+        let version_regexp = "dosbox-staging-[a-zA-Z0-9-]*-([0-9]+\.[0-9]+\.[0-9]+.+)"
+        let release = data.artifacts.find(a => a.name.match(version_regexp))
```
Comment (weirddan455) [inline website/docs/releases/development-builds.md:46]: Just gonna go ahead and trust you on that one :laughing:

**Review by weirddan455 — APPROVED** [review]
Looks good.  A few small concerns about how robust invoking this shell script will be for CI but it looks a lot simpler than what we had so I think it's a positive change.

File: .github/workflows/linux.yml (hunk 2 above)
Comment (johnnovak) [reply .github/workflows/linux.yml:375]: The antivirus scan in the Windows workflow added to the build time significantly. We decided with @kklobe that it has to go as the risk of virus-infection is extremely low at the CI-loop level.

So I thought OK, let's be consistent here and remove it from the Linux workflow as well.

Not sure how much this added to the total run time.

File: .github/workflows/linux.yml (hunk 3 above)
Comment (johnnovak) [reply .github/workflows/linux.yml:393]: Yeah, plus to generate a changelog that probably no-one ever looked at. I certainly did not know it even existed until I started looking into these workflows more deeply 😆

File: website/docs/releases/development-builds.md (hunk 4 above)
Comment (johnnovak) [reply website/docs/releases/development-builds.md:46]: > Just gonna go ahead and trust you on that one 😆

Yeah, tested it for a good 5 minutes with a variety inputs and it worked fine. 

Regexps are kinda "write-only" 😆

File: meson.build (hunk 1 above)
Comment (johnnovak) [reply meson.build]: 1. We could make that work either via WSL (Windows 10 is our minimum requirement anyway), or alternatively a batch file as you suggested. It seems solvable, so we'll cross that bridge when we get there. 

2. Good tip. Changed and also improved the path handling for running the tests (see last commit).

File: src/config.h.in [hunk 5]
```diff
@@ -21,30 +21,14 @@
...
-
-// Name of the emulator
-#define DOSBOX_NAME "DOSBox Staging"
-// Development team name
-#define DOSBOX_TEAM "The " DOSBOX_NAME " Team"
-// Copyright string
-#define DOSBOX_COPYRIGHT "(C) " DOSBOX_TEAM
+// This file is a template for the Meson build process.
+//
+// '@name@' placeholders are substituted with concrete values at the start of
```
Comment (kklobe) [inline src/config.h.in:26]: this throws a warning during build:

```log
../../meson.build:1296: WARNING: The variable(s) 'name' in the input file 'src/config.h.in' are not present in the given configuration data.
```

maybe escape it, e.g.:

```cpp
'@''name@'
```

Comment (johnnovak) [reply src/config.h.in]: Rephrased it a bit, now the warning is gone.

**Pull request merged**